### ISSUE TITLE
Seperate grad norm computation from `torch.nn.utils.clip_grad_norm_`

### ISSUE THREAD

**Issue by zijian-hu** [pytorch/pytorch#139467]
### 🚀 The feature, motivation and pitch

Gradient norm clipping requires computing the total gradient norm across the entire model first. The current design of `torch.nn.utils.clip_grad_norm_` is insufficient for cases like pipeline parallelism (PP).

When using pipeline parallelism (PP), the grad norm need to be computed on each PP stage and then reduced across all PP stages. Separating the grad norm computation from `torch.nn.utils.clip_grad_norm_` would allow developers to properly reduce the grad norm before clipping.

### Additional context

See https://github.com/pytorch/torchtitan/issues/596 and https://github.com/pytorch/torchtitan/pull/649 for more context.

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by awgu** [comment]
We discussed this internally with @janeyx99. There are two separate issues:
1. Provide an op for total norm computation and include a fused op for it (like https://github.com/pytorch/pytorch/pull/134327/)
2. Separate `torch.nn.utils.clip_grad_norm_` to a top half that calls the op from 1 and then a bottom half that applies the clipping

Distributed code will benefit from both 1 and possibly also the separation in 2 (to avoid code duplication).

**Comment by mikaylagawarecki** [comment]
If we split clip_grad_norm_ into `nn.utils.grad_norm` and `nn.utils.clip` (or something of that flavor) does that help? Would it be more useful if these live at the aten level (so fast kernels can be registered) or at the python level?

**Comment by awgu** [comment]
- aten op for total norm would be helpful (can be used for other tensors, not only for grads, so may be better than `nn.utils.grad_norm`)
- clipping can be at Python level

**Comment by mikaylagawarecki** [comment]
From discussion with @wconstab a python op for computing norm and a python op for scaling should be sufficient for the PP use case (without fast kernels) opened https://github.com/pytorch/pytorch/pull/139662 to address this

`foreach_global_norm` is a separate issue and per @janeyx99 likely will not land in core.

**Comment by janeyx99** [comment]
Or that PR as is will not land, but I would not object a cleaned up version of it with clearer semantics (+ being out of foreach). For example, the norm part could be confused as a norm over all the values in the Tensors of the List vs a norm of norms (what it is)

**Comment by zijian-hu** [comment]
Great job! I really like the new design!
